Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,592 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#import <Starboard.h>
+#import <StubReturn.h>
+#import <Foundation/NSDecimal.h>
+
+#define IS_NEGATIVE(decimal) (decimal)->_isNegative
+
+#define IS_POSITIVE(decimal) !IS_NEGATIVE(decimal)
+
+#define IS_ZERO(decimal) (((decimal)->_length == 0) && IS_POSITIVE(decimal))
+
+#define INVALIDATE_DECIMAL(decimal) \
+    \
+decimal->_length = 0;               \
+    \
+decimal->_isNegative = YES;
+
+#define SET_TO_ZERO(decimal) \
+    \
+decimal->_length = 0;        \
+    \
+decimal->_isNegative = NO;   \
+    \
+NSDecimalCompact(decimal);
+
+#define VALIDATE_ARITHMETIC_INPUT(result, leftOperand, rightOperand)                 \
+    if (NSDecimalIsNotANumber(leftOperand) || NSDecimalIsNotANumber(rightOperand)) { \
+        INVALIDATE_DECIMAL(result);                                                  \
+        return NSCalculationOverflow;                                                \
+    }
+
+#define REMINDER(decimal) (decimal->_remainder)
+
+typedef struct {
+    NSDecimal decimal;
+    unsigned short* _mantissa;
+    unsigned short _remainder;
+    // only used for cases where we have carryover
+    unsigned int _extendedLen;
+} NSDecimalInternal;
+
+/**
+ @Status Interoperable
+*/
+void NSDecimalCopy(NSDecimal* destination, const NSDecimal* source) {
+    if (!source || !destination) {
+        return;
+    }
+
+    destination->_isNegative = source->_isNegative;
+    destination->_exponent = source->_exponent;
+    destination->_length = source->_length;
+    destination->_isCompact = source->_isCompact;
+    destination->_reserved = source->_reserved;
+    memcpy(&destination->_mantissa, &source->_mantissa, NSDecimalMaxSize * sizeof(unsigned short));
+}
+
+/**
+ @Status Interoperable
+*/
+BOOL NSDecimalIsNotANumber(const NSDecimal* dcm) {
+    THROW_NS_IF_FALSE(E_INVALIDARG, dcm != NULL);
+    return !dcm->_length && IS_NEGATIVE(dcm);
+}
+
+void _InternalDecimalCopy(NSDecimal* destination, NSDecimalInternal* source) {
+    if (!source || !destination) {
+        return;
+    }
+    NSDecimalCopy(destination, &(source->decimal));
+    memcpy(destination->_mantissa, source->_mantissa, NSDecimalMaxSize * sizeof(unsigned short));
+}
+
+NSCalculationError _subtraction(NSDecimal* result, const NSDecimal* left, const NSDecimal* right) {
```

Comment:
In general, assumptions about parameters going into each function, and abstract description of each major step in the function would be good. (I at this moment am still parsing why this function is broken into two steps, and a comment for each step would've helped greatly)

---
In reply to: [80536966](https://github.com/Microsoft/WinObjC/pull/1032#discussion_r80536966) [](ancestors = 80536966)